control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go

```diff
@@ -2862,6 +2869,72 @@ func (r *reconciler) reconcileAzureCloudNodeManager(ctx context.Context, image s
 	return errs
 }
 
+// reconcileGlobalPullSecret reconciles the original pull secret given by HCP and merges it with a new pull secret provided by the user.
+// The new pull secret is only stored in the DataPlane side so, it's not exposed in the API. It lives in the kube-system namespace of the DataPlane.
+// If that PS exists, the HCCO deploys a DaemonSet which mounts the whole Root FS of the node, and merges the new PS with the original one.
+// If the PS doesn't exist, the HCCO doesn't do anything.
+func (r *reconciler) reconcileGlobalPullSecret(ctx context.Context, hcp *hyperv1.HostedControlPlane, cpoImage string) error {
+	var (
+		additionalPullSecretBytes []byte
+		originalPullSecretBytes   []byte
+		globalPullSecretBytes     []byte
+		err                       error
+	)
+	log := ctrl.LoggerFrom(ctx)
+
+	additionalPullSecret := manifests.UserProvidedPullSecret(UserProvidedPullSecretNamespace)
+	if err := r.uncachedClient.Get(ctx, client.ObjectKeyFromObject(additionalPullSecret), additionalPullSecret); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get user provided pull secret: %w", err)
+	}
+
+	// If the PS doesn't exist, the HCCO doesn't do anything.
+	if additionalPullSecret.Data == nil {
+		return nil
+	}
+
+	if additionalPullSecretBytes, err = globalpullsecret.ValidateAdditionalPullSecret(additionalPullSecret); err != nil {
+		return fmt.Errorf("failed to validate additional pull secret: %w", err)
+	}
+
+	log.Info("Valid additional pull secret found in the DataPlane, reconciling global pull secret")
+
+	// Get the original pull secret
+	originalPullSecret := manifests.PullSecret(hcp.Namespace)
+	if err := r.cpClient.Get(ctx, client.ObjectKeyFromObject(originalPullSecret), originalPullSecret); err != nil {
+		return fmt.Errorf("failed to get original pull secret: %w", err)
+	}
+
+	// Asumming hcp pull secret is valid
+	originalPullSecretBytes = originalPullSecret.Data[corev1.DockerConfigJsonKey]
+
+	// Merge the additional pull secret with the original pull secret
+	if globalPullSecretBytes, err = globalpullsecret.MergePullSecrets(originalPullSecretBytes, additionalPullSecretBytes); err != nil {
+		return fmt.Errorf("failed to merge pull secrets: %w", err)
+	}
+
+	// Create secret in the DataPlane
+	secret := manifests.GlobalPullSecret(UserProvidedPullSecretNamespace)
+	if _, err := r.CreateOrUpdate(ctx, r.uncachedClient, secret, func() error {
+		secret.Data = map[string][]byte{
+			corev1.DockerConfigJsonKey: globalPullSecretBytes,
```

Comment:
what happens if someone modifies the globalPullSecret directly?

Reply:
Well, it depends on the timing, but we expect that could happen 2 things:

- The daemonSet will detects the change
- The daemonSet changes the config.json in the nodes
- The HCCO detects that the globalPullSecret is not well set and overrides it
- The daemonSet detects the change
- The daemonSet changes the config.json in the nodes

Or 
- The HCCO detects that the globalPullSecret is not well set and overrides it
- The daemonSet detects the change
- The daemonSet changes the config.json in the nodes